### DISCUSSION TITLE
Pod Scheduling - How to get pods to distribute across cluster?

### DISCUSSION THREAD

**Discussion by jlphillipsphd** [longhorn/longhorn#4877] [Q&A] [answered]
I have been running deployments of JupyerHub (https://z2jh.jupyter.org/en/stable/) on my cluster with Longhorn providing storage for user pods.

Normally, I have been using the default k8s scheduler to spread those pods evenly across my k3s cluster (I am not using any sort of autoscaling). This used to work well with Longhorn.

However, recently I have noticed that all of my pods are now aggregating onto one node more than the others in the cluster. I don't really know if Longhorn is the root cause of this or not, but I don't have any other components that I can think would be causing this... Things seemed to stay fairly well-distributed for a couple of months, but now the user pods are all trying to load up on one node until that node is full. (Eventually the scheduler will start to fill up the other nodes, but we aren't using the cluster resources very well when this is case.) There are 3 nodes, each volume has 3 replicas, with one replica on each node, so I can't see why this would be happening. Is there something I can try to ensure that Longhorn is willing to equally prefer any replica on the cluster? Does this even seem like something Longhorn would be trying to do?

**Comment by jlphillipsphd** [comment, marked as answer]
I think I just tracked down the issue in case anyone comes across this in the future...

I looked through the UI and found the instance-manager list, noticed that I had some older instance manager engines still running after the last upgrade cycle a month or so ago. Those engines had just not reached a reference count of zero and so were never removed due to some long-running workloads. This was also true on -both- of the nodes that the scheduler was avoiding. The instance managers request a decent-enough chunk of cpu/memory such that the scheduler could see that as a need to prefer the remaining node.

I remember reading in the documentation how there could be old instance managers left after upgrade, and I just didn't want to stop any workload at the time to avoid downtime. (https://longhorn.io/kb/troubleshooting-some-old-instance-manager-pods-are-still-running-after-upgrade/) I thought this would eventually work out over time, but a few long-running workloads were preventing this process from completing in it's entirety.

To fix it, I rescaled the deployments involved down to zero replicas (pods) and the back up to the usual number of replicas - after this the references for the impacted volumes moved to the newer engines. Once those engines finally exited, I saw the node cpu/memory requests now reduce to the same as the other node. Then, the next thing I saw is the reference counts start getting distributed across the cluster engines more evenly and my users are being spread out across the cluster once again.

**Comment by shuo-wu** [comment]
Checked this with Longhorn team members, we don't think there is a preference during the pod scheduling if all worker nodes have enough resources to pick up the pod. In other words, only when some of the nodes have the old engine instance manager pods and are not capable of taking the workload pods that requests too many CPU/memory resources, the scheduling bias will happen.
